core/src/test/java/org/apache/calcite/rex/RexBuilderTest.java

```diff
@@ -558,6 +566,101 @@ private void checkDate(RexNode node) {
     checkBigDecimalLiteral(builder, "-73786976294838206464");
   }
 
+  /** Tests {@link RexCopier#visitOver(RexOver)} */
+  @Test public void testCopyOver() {
+    final RelDataTypeFactory typeFactory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+    final RexBuilder builder = new RexBuilder(typeFactory);
+    final RelDataType type = typeFactory.createSqlType(SqlTypeName.INTEGER);
+    final RexOver node = (RexOver) builder.makeOver(type,
+        SqlStdOperatorTable.COUNT,
+        ImmutableList.of(builder.makeInputRef(type, 0)),
+        ImmutableList.of(builder.makeInputRef(type, 1)),
+        ImmutableList.of(
+            new RexFieldCollation(
+                builder.makeInputRef(type, 2), ImmutableSet.of())),
+        RexWindowBound.create(
+            SqlWindow.createUnboundedPreceding(SqlParserPos.ZERO), null),
+        RexWindowBound.create(
+            SqlWindow.createCurrentRow(SqlParserPos.ZERO), null),
+        true, true, false, false, false);
+    final RexNode copy = builder.copy(node);
```

Comment:
As far as I understand the code, the main advantage of RexCopier (over RexShuttle) is that it can copy a RexNode copying also the type from one factory to the other. The tests are good but they do not seem to verify that the typeFactory is used correctly.

Reply:
Makes sense. I will try to add more accurate tests. Thank you for raising this.

Reply:
I like the idea of `MySqlTypeFactoryImpl` ;) Since the factory is passed in the RexBuilder maybe it would look more realistic if you used a different builders for building and copying the expression. I leave the decision up to you!

Reply:
Thanks. I prefer to keep it this way for saving some time~~